### PR TITLE
Purge resources on rasterizer teardown

### PR DESCRIPTION
**Pull request by dnfield**
This should normally happen when the GrDirectContext is collected, but for various reasons that may not happen or it may take longer to achieve. Roughly speaking, this can happen in an Android app that:

- Renders some scene that causes resource cache occupation
- Gets dismissed in a state where the resource cache has purgeable bytes
- Preserves the Flutter Engine/Dart VM for some time after that, in some way where the GrDirectContext does not get collected

Adds a test that causes the Skia resource cache to become occupied, and then asserts that it has purgeable bytes before teardown and no purgeable bytes after teardown.

Fixes https://github.com/flutter/flutter/issues/105453 - the context may still be retained, but the app will no longer suffer from retaining purgeable cache memory.

### CODE REVIEW

File: shell/common/rasterizer_unittests.cc [hunk 1]
```diff
@@ -76,7 +77,7 @@ class MockExternalViewEmbedder : public ExternalViewEmbedder {
 }  // namespace
 
 TEST(RasterizerTest, create) {
-  MockDelegate delegate;
+  NiceMock<MockDelegate> delegate;
```
Comment (dnfield) [inline shell/common/rasterizer_unittests.cc:80]: These changes are not essential to this patch, but eliminate a lot of noise when running these test targets related to unexpected mock function calls.

**Comment by dnfield** [comment]
@goderbauer FYI - not sure what impact this might have on multi-window

**Comment by dnfield** [comment]
(looking into test failures...)

**Comment by dnfield** [comment]
The ShellTest vulkan related tests are crashing because they can't find libvulkan.so - it's not part of this build and isn't available on CI AFAICT.

They pass before this change because no calls are made that actually touch the rendering backend, but now tearing down the rasterizer may trigger such calls.

**Comment by dnfield** [comment]
I'm not sure about the FEMU failures. The test that's timing out is ShellTest.ReportTimingsIsCalled. I do see in the log that `libvulkan.so` is failing to transfer over to the device - but I wouldn't expect that to make this particular test fail. See https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8811953693866211137/+/u/FEMU_Test/test:_Run_FEMU_Test_Suite_shell_tests/stdout#L34492_0

@akbiggs do you have any ideas about how to debug this further? I do see something in the test file about tests with similar names being skipped (but those tests don't exist anymore). The test is just timing out and getting killed. One other possibility is that the new call in rasterizer teardown is somehow deadlocking...

File: shell/common/rasterizer.cc [hunk 2]
```diff
@@ -87,10 +87,14 @@ void Rasterizer::TeardownExternalViewEmbedder() {
...
+    auto context_switch = surface_->MakeRenderContextCurrent();
+    if (context_switch->GetResult()) {
+      compositor_context_->OnGrContextDestroyed();
+      if (auto* context = surface_->GetContext()) {
+        context->purgeUnlockedResources(/*scratchResourcesOnly=*/false);
+      }
+    }
   }
 
   surface_.reset();
```
Comment (jason-simmons) [inline shell/common/rasterizer.cc:100]: nit: can move the `surface_.reset()` into the `if (surface_)` block

**Review by jason-simmons — APPROVED** [review]

File: shell/common/rasterizer.cc (hunk 2 above)
Comment (dnfield) [reply shell/common/rasterizer.cc]: Done

**Comment by dnfield** [comment]
FEMU failures seem to be related to trying to use swiftshader on FEMU. Using system libvulkan seems to be making more progress... here's hoping.

File: shell/common/shell_test_platform_view_vulkan.cc [hunk 3]
```diff
@@ -8,6 +8,16 @@
 #include "flutter/shell/common/context_options.h"
 #include "flutter/vulkan/vulkan_utilities.h"
 
+#if OS_FUCHSIA
+#define VULKAN_SO_PATH "libvulkan.so"
```
Comment (dnfield) [inline shell/common/shell_test_platform_view_vulkan.cc:12]: @akbiggs FYI

**Pull request merged**